Regex lookahead (#439)
arcus/education_modules: regular_expressions_lookaheads/regular_expressions_lookaheads.md

<!--

author:   Joy Payton
email:    paytonk@chop.edu
version:  1.0.0
current_version_description: Initial version
module_type: standard
docs_version: 1.0.0
language: en
narrator: US English Female
mode: Textbook

> previous_sequential_module: regular_expressions_basics

Reviewer: ```suggestion
```

@end

@learning_objectives  
After completion of this module, learners will be able to:

- Explain the difference between "moving" ahead and "looking" ahead in regular expression parsing
- Explain why a "lookahead" can be useful in a regular expression

@end
good_first_module: false
coding_required: true
coding_level: intermediate
sequence_name: regex
previous_sequential_module: regular_expressions_anchors_boundaries
@sets_you_up_for

@end 
@depends_on_knowledge_available_in
- demystifying_regular_expressions
- regular_expressions_basics
- regular_expressions_groups

> - regular_expressions_anchors_boundaries

Reviewer: ```suggestion
```
I think it doesn't actually depend on anything in the anchors/boundaries module?

A good place to practice regular expressions is an online regex testing website.  We like **Regex 101**, at https://www.regex101.com.

Regex consists of **tokens** which represent characters that appear in a particular location, and **quantifiers**, which give the number of times a given token repeats.

- **Tokens** include `\d`, `[A-Za-z]`, `A`, and `.`

- **Quantifiers** include `*`, `{2,5}`, and `+`

Regular expressions can make use of **groups**, both **capturing** and **non-capturing**.

If this refresher doesn't seem familiar, you might want to consider reviewing [Demystifying Regular Expressions](https://liascript.github.io/course/?https://raw.githubusercontent.com/arcus/education_modules/main/demystifying_regular_expressions/demystifying_regular_expressions.md#1), [Regular Expressions Basics](https://liascript.github.io/course/?https://raw.githubusercontent.com/arcus/education_modules/main/regular_expressions_basics/regular_expressions_basics.md#1), and [Regular Expressions: Groups](https://liascript.github.io/course/?https://raw.githubusercontent.com/arcus/education_modules/main/regular_expressions_groups/regular_expressions_groups.md#1).

## Lookaheads

> We consider this module to be intermediate, because it assumes basic knowledge covers more advanced regular expressions topics, without delving into the advanced topic of using regular expressions within the context of code.  

Reviewer: ```suggestion
We consider this module to be intermediate, because it assumes basic knowledge and covers more advanced regular expressions topics, without delving into the advanced topic of using regular expressions within the context of code.  
```

A log file includes text like:

`20.3 22.0 19.5 NA NA 30.9 30.2 33.8 29.5 NA NA`.

In your code, you'd like to process each pair of consecutive decimal values to study rate of change, so you are using regular expressions to return consecutive pairs of decimal numbers.  You only want consecutive pairs of numbers, not any consecutive pairs that have `NA` values.

So, you'd expect five pairs:

* The pair 20.3 and 22.0
* The pair 22.0 and 19.5
* The pair 30.9 and 30.2
* The pair 30.2 and 33.8
* The pair 33.8 and 29.5

You start by writing this regular expression which uses capture groups to extract consecutive decimal numbers: `([\d\.]+)\s+([\d\.]+)`.  You [try this out in your favorite regex checker](https://regex101.com/r/pecVbn/1), and you're confused! Why do you only capture three pairs of numbers?  There should be five pairs!

![The regular expression shows three matches, each of which has two captured groups.](media/moveahead.png) <!-- style = "border: 1px solid rgb(var(--color-highlight)); " -->

> See if you can figure this out before advancing to the next page.

Reviewer: ```suggestion
See if you can figure out why only three pairs, not all five, were found to be matches before advancing to the next page.
```

This continues and we manage to find all 5 pairs of consecutive numbers, even when the patterns overlap.  

The **lookahead** regular expression syntax takes the form of `(?=regex)` where `regex` is replaced with what you're looking for.  So, for example, our pattern for the second decimal number, which is currently in a capturing group and looks like: `([\d\.]+)` could be placed between `(?=` and `)`, giving us `(?=([\d\.]+))`.
Go ahead and [try that in a regex checker](https://regex101.com/r/NvzPb0/1) and you should see something like the following. 

![The regular expression shows five matches, each of which has two captured groups.](media/lookahead.png) <!-- style = "border: 1px solid rgb(var(--color-highlight));" -->

In the image above, the overlapping color in the "Text String" box can make the visual interpretation of the image tricky.  Look over in the "Match Information" section to see all five matches!

> ## 

Reviewer: ```suggestion
```

<div class = "behind-the-scenes">
<b style="color: rgb(var(--color-highlight));">Behind the scenes</b><br>

There are other types of "lookarounds," including **lookbehinds**, that you might consider learning if you write a lot of regular expressions, but for now, the lookahead by itself will serve you well for more complex cases.

</div>

> ## Quiz 4: Anchors, Boundaries, and Lookaheads

Reviewer: ```suggestion
## Quiz: Lookaheads
```
